File: go/0017M-电话号码的字母组合.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-12 16:59:55
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func letterCombinations(digits string) []string {
	letter := [10]string{
		"",     // 0
		"",     // 1
		"abc",  // 2
		"def",  // 3
		"ghi",  // 4
		"jkl",  // 5
		"mno",  // 6
		"pqrs", // 7
		"tuv",  // 8
		"wxyz", // 9
	}
	ans := make([]string, 0)
	if len(digits) == 0 {
		return ans
	}
	str := ""
	var dfs func(p int)
	dfs = func(p int) {
		if len(str) == len(digits) {
			ans = append(ans, str)
			return
		}
		digit := digits[p] - '0'
		letters := letter[digit]
		for _, v := range letters {
			str += string(v)
			dfs(p + 1) // 这里 digits 右移
			str = str[:len(str)-1]
		}
	}
	dfs(0)
	return ans
}

File: cpp/0017M-电话号码的字母组合.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-26 14:13:15
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    string letter[10] = {
        "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz",
    };

    vector<string> ans;
    string tmp;

    void dfs(string digits, int idx) {
        if (tmp.size() == digits.size()) {
            ans.push_back(tmp);
            return;
        }
        string l = letter[digits[idx] - '0'];
        for (char c : l) {
            tmp.push_back(c);
            dfs(digits, idx + 1);
            tmp.pop_back();
        }
    }

    vector<string> letterCombinations(string digits) {
        if (digits == "") return {};
        dfs(digits, 0);
        return ans;
    }
};